src/MiniExec.sol

```diff
@@ -31,62 +35,37 @@ contract MiniExecImplementation is IZkEVMBridgeMessageReceiver {
             }
         }
     }
-
-    function _metadata() internal view returns (address, uint32) {
-        return creator.metadata(address(this));
-    }
 }
 
-contract MiniExecProxy is Proxy {
+contract MiniExecProxy is Proxy, MiniExecStorage {
     error OnlySelf();
 
     MiniExecFactory immutable creator;
 
-    constructor() {
-        creator = MiniExecFactory(msg.sender);
+    constructor(address owner, uint32 networkId, address impl) {
+        remoteOwner = owner;
+        remoteNetwork = networkId;
+        implementation = impl;
     }
 
-    function updateImplementation(address _newImplementation) external onlySelf {
-        creator.setImplementation(_newImplementation);
-    }
-
-    function _implementation() internal view override returns (address) {
-        return creator.implementations(address(this));
-    }
-
-    modifier onlySelf() {
+    function updateImplementation(address _newImplementation) external {
         if (msg.sender != address(this)) revert OnlySelf();
-        _;
+        implementation = _newImplementation;
     }
 
-    receive() external payable {
-        _fallback();
+    function _implementation() internal view override returns (address) {
+        return implementation;
     }
 }
 
 contract MiniExecFactory {
-    struct Metadata {
-        address owner;
-        uint32 networkId;
-    }
-
-    mapping(address => Metadata) public metadata;
-    mapping(address => address) public implementations;
-
     MiniExecImplementation public immutable miniExecImplementation;
```

Comment:
hmm just we also have a way to change this implementation?

Reply:
added